README.md
# Secure File Host
> Filesystem host toolkit for Browser-Desktop datasource connections

[![Build Status](https://travis-ci.org/buttercup/secure-file-host.svg?branch=master)](https://travis-ci.org/buttercup/secure-file-host)

## About

This library provides an **event emitter** and an **Express application** for use with the [Buttercup Desktop application](https://github.com/buttercup/buttercup-desktop) - It's designed to provide an interface for [Browser extension](https://github.com/buttercup/buttercup-browser-extension) users to connect on for reading and writing vault files on the local system.

Some users may not want to have their vault files synchronised via a cloud storage service, and this tool is designed to allow them to read files on their PC without the need for connecting to external services.

## Installation

Simply run `npm install @buttercup/secure-file-host --save` to install.

This library is only designed to work with Node JS.

## Usage

Create and run an express server:

```javascript
const { startFileHost } = require("@buttercup/secure-file-host");

const host = startFileHost(9998);
// host.key is the encryption key
```

_In this example, the browser can connect to the file host service on port `9998`._

For repeated use, you may want to provide your own key:

```javascript
startFileHost(9998, "some-encryption-key");
```

You can stop the service when closing the application:

```javascript
const host = startFileHost(9998);

// later
host.stop();
```

If a user wants to cancel the handskahe procedure, make sure to call the `host.cancel()` method to prevent a deadlock.

More details available in the [API documentation](API.md)

### Debugging

You can run a debug server that uses the host by running `npm start`. You can stop the server by pressing `Ctrl+C`.

### Emitted events

The returned `host` object contains an `emitter` property, which is an event emitter instance. It emits the following events:

| Event Slug    | Schedule               | Description                                               |
|---------------|------------------------|-----------------------------------------------------------|
| `codeReady`   | On client connect      | A client has connected and is awaiting a code to be presented so that the **user** can enter it in the other application's interface. |
| `connectionAvailabilityChanged` | When the status of the host in terms of it accepting new connections changes | Called with an object like `{ available: true }` when the availability of the host changes. It will be fired with `{ available: false }` when someone is trying to connect, and it will stay unavailable until it is either connected or cancelled. If a wrong code is attempted it will stay unavailable for a further 15 seconds. |
| `connected`   |  When connection and handshake procedures have been completed | When the entire handshake procedure has been completed successfully the `connected` event is fired with no arguments. |
| `connectionFailed` | When an incorrect code is entered | Fired when a client tries to use an incorrect code. |


CHANGELOG.md
# Secure File Host changelog

## v0.3.0
_2021-03-09_

 * Dependency updates

## v0.2.1
_2018-11-24_

 * **Bugfix**: `/get/directory` would fail on Windows

## v0.2.0
_2018-11-19_

 * Filter out dot-prefixed _hidden_ files and folders

## v0.1.3
_2018-11-14_

 * Add `/put/file` API method

## v0.1.2
_2018-11-14_

 * Add `connectionFailed` event

## v0.1.1
_2018-11-14_

 * Add `host.cancel()` method

## v0.1.0
_2018-11-04_

 * Initial release


test/app.js
const { createHost } = require("../source/host.js");

function getNewApp() {
    return createHost(null, "testing");
}

module.exports = {
    getNewApp
};


test/standalone.js
const fs = require("fs");
const path = require("path");
const chalk = require("chalk");
const { startFileHost } = require("../source/index.js");

const KEYPATH = path.resolve(__dirname, "../standalone.key");
const PORT = 12821;

function error(...args) {
    console.log(chalk.red("[STANDALONE]"), ...args);
}
function log(...args) {
    console.log(chalk.green("[STANDALONE]"), ...args);
}

log("Starting standalone server");
let key,
    fileWasRead = false;
try {
    key = fs.readFileSync(KEYPATH, "utf8");
    fileWasRead = true;
} catch (err) {
    error(`No key at path: ${KEYPATH}`);
}
if (!key && fileWasRead) {
    throw new Error("Key from key-file was invalid");
}

const host = startFileHost(PORT, key);
if (!host.key) {
    throw new Error("Key from host was invalid");
}
fs.writeFileSync(KEYPATH, host.key);
host.emitter.on("codeReady", info => {
    log(`Code ready: ${info.code}`);
});
host.emitter.on("connected", () => {
    log("Client connected with code");
});
host.emitter.on("connectionFailed", () => {
    error("Client connection failed");
});
host.emitter.on("connectionAvailabilityChanged", info => {
    log(`Connection availability changed: ${info.available ? "Available" : "Unavailable"}`);
});
log(`Started server on port ${PORT} with key: ${host.key}`);

process.on("SIGINT", function() {
    log("Interrupt request detected");
    host.stop();
    process.exit(0);
});


test/specs/api.spec.js
const path = require("path");
const fs = require("fs");
const request = require("supertest");
const pify = require("pify");
const getHomeDirectory = require("home-dir");
const rimrafC = require("rimraf");
const { getNewApp } = require("../app.js");
const { decryptString, encryptString } = require("../../source/crypto.js");

const rimraf = pify(rimrafC);
const readFile = pify(fs.readFile);

describe("API", function() {
    let host;

    beforeEach(function() {
        host = getNewApp();
    });

    afterEach(function() {
        host.stop();
    });

    describe("GET /", function() {
        it("returns expected properties", done => {
            request(host.app)
                .get("/")
                .expect("Content-Type", /application\/json/)
                .expect(200)
                .end((err, res) => {
                    if (err) return done(err);
                    expect(res.body).to.have.property("type", "secure-file-host");
                    expect(res.body).to.have.property("status", "ok");
                    expect(res.body).to.have.property("ready", true);
                    done();
                });
        });
    });

    describe("GET /connect", function() {
        it("responds with OK status", done => {
            request(host.app)
                .get("/connect")
                .expect("Content-Type", /application\/json/)
                .expect(200)
                .end((err, res) => {
                    if (err) return done(err);
                    expect(res.body).to.have.property("status", "ok");
                    done();
                });
        });

        it("responds with 503 Unavailable if called twice", done => {
            request(host.app)
                .get("/connect")
                .expect("Content-Type", /application\/json/)
                .expect(200)
                .end((err, res) => {
                    if (err) return done(err);
                    request(host.app)
                        .get("/connect")
                        .expect("Content-Type", /text\/plain/)
                        .expect(503)
                        .end((err, res) => {
                            if (err) return done(err);
                            done();
                        });
                });
        });

        it("emits 'codeReady' with connection code", done => {
            host.emitter.once("codeReady", data => {
                expect(data).to.have.property("code")
                    .that.is.a("string")
                    .that.matches(/^[A-Z0-9]{6}$/);
                done();
            });
            request(host.app)
                .get("/connect")
                .end((err, res) => {
                    if (err) return done(err);
                });
        });
    });

    describe("GET /connect/(code)", function() {
        it("responds with 400 Bad Request if no code sequence has been started", done => {
            request(host.app)
                .get("/connect/ABC123")
                .expect("Content-Type", /text\/plain/)
                .expect(400)
                .end((err, res) => {
                    if (err) return done(err);
                    done();
                });
        });

        it("emits 'connectionFailed' when wrong code entered", done => {
            host.emitter.once("connectionFailed", done);
            request(host.app)
                .get("/connect/ABC123")
                .end((err, res) => {
                    if (err) return done(err);
                });
        });

        it("responds with an encrypted code when prepared", done => {
            let code;
            host.emitter.once("codeReady", result => {
                code = result.code;
            });
            request(host.app)
                .get("/connect")
                .expect("Content-Type", /application\/json/)
                .expect(200)
                .end((err, res) => {
                    if (err) return done(err);
                    request(host.app)
                        .get(`/connect/${code}`)
                        .expect("Content-Type", /application\/json/)
                        .expect(200)
                        .end((err, res) => {
                            if (err) return done(err);
                            expect(res.body).to.have.property("status", "ok");
                            decryptString(res.body.payload, code)
                                .then(result => {
                                    expect(result).to.equal("testing");
                                    done();
                                })
                                .catch(done);
                        });
                });
        });
    });

    describe("POST /get/directory", function() {
        const TARGET_DIR = path.resolve(__dirname, "../assets");
        let encryptedPayload;

        beforeEach(function() {
            return encryptString(TARGET_DIR, "testing")
                .then(encrypted => {
                    encryptedPayload = encrypted;
                });
        });

        it("returns expected contents", done => {
            request(host.app)
                .post("/get/directory")
                .send({
                    payload: encryptedPayload
                })
                .expect("Content-Type", /application\/json/)
                .expect(200)
                .end((err, res) => {
                    if (err) return done(err);
                    expect(res.body).to.have.property("status", "ok");
                    decryptString(res.body.payload, "testing")
                        .then(filesRaw => {
                            const files = JSON.parse(filesRaw);
                            expect(files).to.have.lengthOf(1);
                            expect(files[0]).to.have.property("name", "getfile.txt");
                            expect(files[0]).to.have.property("type", "file");
                            done();
                        })
                        .catch(done);
                });
        });

        it("responds with 401 Unauthorized if decryption of the payload fails", done => {
            request(host.app)
                .post("/get/directory")
                .send({
                    payload: "bad-payload"
                })
                .expect("Content-Type", /text\/plain/)
                .expect(401)
                .end((err, res) => {
                    if (err) return done(err);
                    done();
                });
        });
    });

    describe("POST /get/file", function() {
        const TARGET_FILE = path.resolve(__dirname, "../assets/getfile.txt");
        let encryptedPayload;

        beforeEach(function() {
            return encryptString(TARGET_FILE, "testing")
                .then(encrypted => {
                    encryptedPayload = encrypted;
                });
        });

        it("returns expected contents", done => {
            request(host.app)
                .post("/get/file")
                .send({
                    payload: encryptedPayload
                })
                .expect("Content-Type", /application\/json/)
                .expect(200)
                .end((err, res) => {
                    if (err) return done(err);
                    expect(res.body).to.have.property("status", "ok");
                    decryptString(res.body.payload, "testing")
                        .then(fileContents => {
                            expect(fileContents.trim()).to.equal("test contents");
                            done();
                        })
                        .catch(done);
                });
        });

        it("responds with 401 Unauthorized if decryption of the payload fails", done => {
            request(host.app)
                .post("/get/file")
                .send({
                    payload: "bad-payload"
                })
                .expect("Content-Type", /text\/plain/)
                .expect(401)
                .end((err, res) => {
                    if (err) return done(err);
                    done();
                });
        });

        it("responds with 404 Not Found when reading a non-existant file", done => {
            encryptString("/not/here.txt", "testing")
                .then(encrypted => {
                    encryptedPayload = encrypted;
                    request(host.app)
                        .post("/get/file")
                        .send({
                            payload: encryptedPayload
                        })
                        .expect("Content-Type", /text\/plain/)
                        .expect(404)
                        .end((err, res) => {
                            if (err) return done(err);
                            done();
                        });
                })
                .catch(done);
        });
    });

    describe("POST /put/file", function() {
        const TARGET_FILE = path.resolve(__dirname, "../assets/putfile.txt");
        let encryptedPayload;

        beforeEach(function() {
            return rimraf(TARGET_FILE)
                .then(() => encryptString(JSON.stringify({
                    filename: TARGET_FILE,
                    contents: "my cool contents"
                }), "testing"))
                .then(encrypted => {
                    encryptedPayload = encrypted;
                });
        });

        afterEach(function() {
            return rimraf(TARGET_FILE);
        });

        it("writes contents of file", done => {
            request(host.app)
                .post("/put/file")
                .send({
                    payload: encryptedPayload
                })
                .expect("Content-Type", /application\/json/)
                .expect(200)
                .end((err, res) => {
                    if (err) return done(err);
                    expect(res.body).to.have.property("status", "ok");
                    readFile(TARGET_FILE, "utf8").then(contents => {
                        expect(contents).to.equal("my cool contents");
                        done();
                    }).catch(done);
                });
        });
    });

    describe("GET /get/homedir", function() {
        it("responds with home directory", done => {
            request(host.app)
                .get("/get/homedir")
                .expect("Content-Type", /application\/json/)
                .expect(200)
                .end((err, res) => {
                    if (err) return done(err);
                    expect(res.body).to.have.property("status", "ok");
                    decryptString(res.body.payload, "testing")
                        .then(homeDir => {
                            expect(homeDir).to.equal(getHomeDirectory());
                            done();
                        })
                        .catch(done);
                });
        });
    });
});


source/code.js
const CIPHER_KEY_CHARACTERS = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789abcdefghijklmnopqrstuvwxyz_-()&*%^#$!@~+=[]{};:,.<>?";
const CIPHER_KEY_LENGTH_MAX = 32;
const CIPHER_KEY_LENGTH_MIN = 24;
const CODE_CHARACTERS = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";

function generateCipherKey() {
    const len = CIPHER_KEY_LENGTH_MIN + Math.floor(Math.random() * (CIPHER_KEY_LENGTH_MAX - CIPHER_KEY_LENGTH_MIN + 1));
    return generateRandomString(len, CIPHER_KEY_CHARACTERS);
}

function generateConnectCode() {
    return generateRandomString(6, CODE_CHARACTERS);
}

function generateRandomString(len, characters) {
    let output = "";
    for (let i = 0; i < len; i += 1) {
        output += characters[Math.floor(Math.random() * characters.length)];
    }
    return output;
}

module.exports = {
    generateCipherKey,
    generateConnectCode
};


source/filesystem.js
const fs = require("fs");
const path = require("path");
const pify = require("pify");
const getHomeDirectory = require("home-dir");

const readDir = pify(fs.readdir);
const readFile = pify(fs.readFile);
const stat = pify(fs.stat);
const writeFile = pify(fs.writeFile);

const FILES_TO_HIDE = /^\./;
const IS_WINDOWS = process.platform === 'win32';

function getDirectoryContents(dir) {
    if (IS_WINDOWS && dir === '/') {
        return handleWindowsRootPath();
    }

    let dirRealPath = normalizePath(dir);
    return readDir(dirRealPath)
        .then(contents => Promise.all(contents.map(filename =>
            stat(path.join(dirRealPath, filename))
                .then(stat => ({
                    name: filename,
                    directory: dir,
                    isFile: stat.isFile(),
                    isDirectory: stat.isDirectory()
                }))
                .catch(err => {
                    console.error(`Error while trying to stat file: ${path.join(dirRealPath, filename)} (${err.message})`);
                    return null;
                })
        )))
        .then(contents =>
            contents
                .filter(item => item && (item.isDirectory || item.isFile) && FILES_TO_HIDE.test(item.name) === false)
                .map(dirEnt => ({
                    name: dirEnt.name,
                    directory: dirEnt.directory,
                    type: dirEnt.isDirectory ? "directory" : "file"
                }))
        );
}

function getFileContents(path) {
    return readFile(normalizePath(path), "utf8");
}

function putFileContents(path, contents) {
    // windows paths must have at least 3 parts due to drive letter
    // 2 or less parts mean script is write to a path that doesn't exist
    if (IS_WINDOWS && path.indexOf('\\') === -1 && path.split('/').length <= 2) {
        return Promise.reject(new Error(`Can't write files on Windows root path`));
    }
    return writeFile(normalizePath(path), contents, "utf8");
}

function getWindowsDrives() {
    return new Promise((resolve, reject) => {
        let stdout = '';
        let stderr;
        const { spawn } = require('child_process'),
            list = spawn('cmd');

        list.stdout.on('data', (data) => stdout += data);
        list.stderr.on('data', (data) => stderr += data);

        list.on('exit', function (code) {
            if (code == 0) {
                let data = stdout.split('\r\n');
                data = data.splice(4, data.length - 7);
                data = data.map(drive => drive.substring(0, 1).toLowerCase());
                resolve(data);
            } else {
                reject({ code, message: stderr });
            }
        });
        list.stdin.write('wmic logicaldisk get caption\n');
        list.stdin.end();
    });

}

function handleWindowsRootPath(dir) {
    return getWindowsDrives()
        .then(drives => {
            return drives.map(drive => {
                return {
                    name: drive,
                    directory: '/',
                    type: 'directory'
                };
            });
        })
        .catch(err => {
            console.error(`Error while trying to get all Windows drives: (${err.message})`);
            return null;
        });
}

// convert unix like path to windows path
function normalizeWindowsPath(dir) {
    // ignore proper windows path
    if (dir.indexOf('\\') > -1) {
        return path.normalize(dir);
    }

    // replace drive letter from /c/ to c:/
    const drive = dir.substring(1, 2) + ':';

    // grab rest of the path
    const filePath = dir.substring(2) || '/';

    // convert path to Windows format
    return path.join(drive, filePath);
}

function normalizePath(dir) {
    return (IS_WINDOWS) ? normalizeWindowsPath(dir) : dir;
}

module.exports = {
    getDirectoryContents,
    getFileContents,
    getHomeDirectory,
    putFileContents
};


source/host.js
const EventEmitter = require("eventemitter3");
const express = require("express");
const bodyParser = require("body-parser");
const pkgInfo = require("../package.json");
const { generateConnectCode } = require("./code.js");
const { decryptString, encryptString } = require("./crypto.js");
const { getDirectoryContents, getFileContents, getHomeDirectory, putFileContents } = require("./filesystem.js");

const RESET_DELAY = 15000;
const SHOW_DURATION = 15000;

/**
 * Host instance
 * @typedef {Object} Host
 * @property {Object} app - ExpressJS application instance
 * @property {Object} emitter - Event emitter instance
 * @property {Object|null} server - ExpressJS server instance
 * @property {Function=} stop - Method to stop the host from serving
 * @property {Function} cancel - Method to cancel the current connection attempt
 */

/**
 * Configure a host instance
 * @param {Host} host The host instance
 * @param {String} key The encryption key
 * @private
 */
function configureHost(host, key) {
    const { app, emitter } = host;
    let connectCode = null,
        busy = false;
    host._timer = null;
    host.cancel = () => {
        clearTimeout(host._timer);
        connectCode = null;
        busy = false;
        emitter.emit("connectionAvailabilityChanged", { available: true });
    };
    const timerReset = () => {
        clearTimeout(host._timer);
        connectCode = null;
        busy = true;
        emitter.emit("connectionFailed");
        emitter.emit("connectionAvailabilityChanged", { available: false });
        host._timer = setTimeout(() => {
            busy = false;
            emitter.emit("connectionAvailabilityChanged", { available: true });
        }, RESET_DELAY);
    };
    app.disable("x-powered-by");
    app.use(bodyParser.json());
    app.use((req, res, next) => {
        res.set("Content-Type", "application/json");
        res.set("Server", `Buttercup/SecureFileHost/${pkgInfo.version}`);
        next();
    });
    app.get("/", (req, res) => {
        res.send(JSON.stringify({
            type: "secure-file-host",
            status: "ok",
            ready: !busy
        }));
    });
    app.get("/connect", (req, res) => {
        if (connectCode || busy) {
            res
                .set("Content-Type", "text/plain")
                .status(503)
                .send("Unavailable");
            return;
        }
        connectCode = generateConnectCode();
        emitter.emit("codeReady", {
            code: connectCode
        });
        busy = true;
        res.send(JSON.stringify({
            status: "ok"
        }));
    });
    app.get("/connect/:code", (req, res) => {
        const { code } = req.params;
        if (connectCode && busy) {
            if (connectCode !== code) {
                timerReset();
                res
                    .set("Content-Type", "text/plain")
                    .status(401)
                    .send("Unauthorized");
                return;
            }
            // Success
            const responseKey = connectCode;
            connectCode = null;
            busy = false;
            emitter.emit("connected");
            encryptString(key, responseKey)
                .then(encrypted => {
                    res.send(JSON.stringify({
                        status: "ok",
                        payload: encrypted
                    }));
                })
                .catch(err => {
                    console.error(err);
                    res
                        .set("Content-Type", "text/plain")
                        .status(500)
                        .send("Internal Server Error");
                });
            return;
        } else if (busy && !connectCode) {
            timerReset();
            res
                .set("Content-Type", "text/plain")
                .status(503)
                .send("Unavailable");
            return;
        }
        timerReset();
        res
            .set("Content-Type", "text/plain")
            .status(400)
            .send("Bad Request");
        return;
    });
    app.post("/get/directory", (req, res) => {
        const { payload } = req.body;
        decryptString(payload, key)
            .then(dir => {
                return getDirectoryContents(dir)
                    .then(items => {
                        return encryptString(JSON.stringify(items), key);
                    })
                    .then(payload => {
                        res.send(JSON.stringify({
                            status: "ok",
                            payload
                        }));
                    })
                    .catch(err => {
                        console.error(err);
                        res
                            .set("Content-Type", "text/plain")
                            .status(500)
                            .send("Internal Server Error");
                    });
            })
            .catch(err => {
                console.error(err);
                res
                    .set("Content-Type", "text/plain")
                    .status(401)
                    .send("Unauthorized");
            });
    });
    app.post("/get/file", (req, res) => {
        const { payload } = req.body;
        decryptString(payload, key)
            .then(filename => {
                return getFileContents(filename)
                    .then(contents => {
                        return encryptString(contents, key);
                    })
                    .then(payload => {
                        res.send(JSON.stringify({
                            status: "ok",
                            payload
                        }));
                    })
                    .catch(err => {
                        console.error(err);
                        if (err.code === "ENOENT") {
                            res
                                .set("Content-Type", "text/plain")
                                .status(404)
                                .send("Not Found");
                            return;
                        }
                        res
                            .set("Content-Type", "text/plain")
                            .status(500)
                            .send("Internal Server Error");
                    });
            })
            .catch(err => {
                console.error(err);
                res
                    .set("Content-Type", "text/plain")
                    .status(401)
                    .send("Unauthorized");
            });
    });
    app.post("/put/file", (req, res) => {
        const { payload } = req.body;
        decryptString(payload, key)
            .then(JSON.parse)
            .then(({ contents, filename }) => {
                return putFileContents(filename, contents)
                    .then(() => {
                        res.send(JSON.stringify({
                            status: "ok"
                        }));
                    })
                    .catch(err => {
                        console.error(err);
                        res
                            .set("Content-Type", "text/plain")
                            .status(500)
                            .send("Internal Server Error");
                    });
            })
            .catch(err => {
                console.error(err);
                res
                    .set("Content-Type", "text/plain")
                    .status(401)
                    .send("Unauthorized");
            });
    });
    app.get("/get/homedir", (req, res) => {
        encryptString(getHomeDirectory(), key)
            .then(payload => {
                res.send(JSON.stringify({
                    status: "ok",
                    payload
                }));
            })
            .catch(err => {
                console.error(err);
                res
                    .set("Content-Type", "text/plain")
                    .status(500)
                    .send("Internal Server Error");
            });
    });
}

/**
 * Create a new host instance
 * @param {Number|undefined} port The port to listen on
 * @param {String} key The encryption key
 * @returns {Host} A new host instance
 */
function createHost(port, key) {
    const app = express();
    const emitter = new EventEmitter();
    const host = {
        app,
        emitter,
        server: null,
        stop: () => stopHost(host)
    };
    configureHost(host, key);
    if (port) {
        host.server = app.listen(port);
    }
    return host;
}

/**
 * Stop a host instance
 * @param {Host} host The host instance
 * @private
 */
function stopHost(host) {
    clearTimeout(host._timer);
    if (host.server) {
        host.server.close();
    }
    delete host.stop;
}

module.exports = {
    createHost
};
